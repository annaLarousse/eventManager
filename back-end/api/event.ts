import fs from 'fs';
import { Express, Request, Response } from 'express';
import { IEvent } from '../models/event';
import { IApiResponse } from '../models/apiResponse';

const dataFilePath: string = __dirname + '/../data/' + 'events.json';
module.exports = (app: Express) => {
  app.get('/events', (_req: Request, res: Response) => {
    res.contentType('application/json');
    const data: string = getEventsFromJSONFile();
    if (!data || !Object.keys(JSON.parse(data)).length)
      res.status(404).json('No events found!');
    else {
      res.status(200);
      res.end(data);
    }
  });

  app.post('/event', (req: Request, res: Response) => {
    res.contentType('application/json');
    let apiResponse: IApiResponse = {
      status: 201,
      message: 'Event added successfully!',
    };
    const newEvent: IEvent = {
      id: -1,
      name: req.body.name ? `${req.body.name}` : '',
      description: req.body.description ? `${req.body.description}` : '',
      startDate: req.body.startDate ? `${req.body.startDate}` : '',
      endDate: req.body.endDate ? `${req.body.endDate}` : '',
    };

    const errorMessage = doValidation(newEvent);
    if (errorMessage) {
      apiResponse = {
        status: 400,
        message: errorMessage,
      };
    }
    if (apiResponse.status !== 400) {
      res.contentType('text/plain');
      let dataString = getEventsFromJSONFile();
      const data = dataString ? JSON.parse(dataString) : {};
      const dataLength: number = Object.keys(data).length;
      const newId: number = dataLength + 1;

      newEvent.id = newId;
      (data[newEvent.id] as Object) = newEvent;

      dataString = JSON.stringify(data, undefined, 2);
      fs.writeFileSync(dataFilePath, dataString);
      apiResponse = {
        status: 201,
        message: 'Event added successfully!',
      };
    }
    res.writeHead(apiResponse.status);
    res.end(JSON.stringify(apiResponse));
  });
};

function getEventsFromJSONFile(): string {
  return fs.readFileSync(dataFilePath, 'utf8');
}

function doValidation(data: IEvent): string | null {
  if (!isValidDate(data.startDate)) {
    return 'Start date must be a valid date.';
  }
  if (!isValidDate(data.endDate)) {
    return 'End date must be a valid date.';
  }
  if (!data.name) {
    return 'Name must be set.';
  }
  if (data.name.length > 32) {
    return 'Event name cannot exceed 32 characters.';
  }
  if (data.endDate <= data.startDate) {
    return 'The end date have to be after the start date.';
  }
  return null;
}

function isValidDate(dateString: string): boolean {
  return !!Date.parse(dateString);
}
