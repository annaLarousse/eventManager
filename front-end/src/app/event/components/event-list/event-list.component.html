<h3>Events list</h3>

<table
  mat-table
  [dataSource]="(events$ | async) || []"
  class="event-list__table mat-elevation-z8"
>
  <!-- Name Column -->
  <ng-container matColumnDef="name">
    <th mat-header-cell *matHeaderCellDef>Name</th>
    <td mat-cell *matCellDef="let element">{{ element.name }}</td>
  </ng-container>

  <!-- Start Date Column -->
  <ng-container matColumnDef="startDate">
    <th mat-header-cell *matHeaderCellDef>Start date</th>
    <td mat-cell *matCellDef="let element">
      {{ element.startDate | date: "medium" }}
    </td>
  </ng-container>

  <!-- End Date Column -->
  <ng-container matColumnDef="endDate">
    <th mat-header-cell *matHeaderCellDef>End date</th>
    <td mat-cell *matCellDef="let element">
      {{ element.endDate | date: "medium" }}
    </td>
  </ng-container>

  <!-- Description Column -->
  <ng-container matColumnDef="description">
    <th mat-header-cell *matHeaderCellDef>Description</th>
    <td mat-cell *matCellDef="let element">{{ element.description }}</td>
  </ng-container>

  <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
  <tr mat-row *matRowDef="let row; columns: displayedColumns"></tr>
</table>
