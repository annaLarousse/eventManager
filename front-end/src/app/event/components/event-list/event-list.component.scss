:host.event-list {
  display: flex;
  flex-direction: column;

  .event-list {
    &__table {
      width: 100%;
      ::ng-deep td {
        &.mat-column-name,
        &.mat-column-startDate,
        &.mat-column-endDate {
          width: 12rem;
        }
      }

      ::ng-deep th.mat-header-cell {
        color: var(--primary);
      }
    }
  }
}
