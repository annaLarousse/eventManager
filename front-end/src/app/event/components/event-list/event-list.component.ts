import { Component } from '@angular/core';
import { Observable } from 'rxjs';
import { IEvent } from '../../../../../../back-end/models/event';
import { EventService } from '../../services/event.service';

@Component({
  selector: 'app-event-list',
  templateUrl: './event-list.component.html',
  styleUrls: ['./event-list.component.scss'],
  host: { class: 'event-list' },
})
export class EventListComponent {
  public events$: Observable<IEvent[]>;

  public displayedColumns: string[] = [
    'name',
    'startDate',
    'endDate',
    'description',
  ];

  constructor(private _eventService: EventService) {
    this.events$ = this._eventService.events$;
  }
}
