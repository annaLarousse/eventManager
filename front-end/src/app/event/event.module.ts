import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { MatTableModule } from '@angular/material/table';
import { EventListComponent } from './components/event-list/event-list.component';
import { EventPageComponent } from './components/event-page/event-page.component';

@NgModule({
  declarations: [EventPageComponent, EventListComponent],
  imports: [CommonModule, HttpClientModule, MatTableModule],
  exports: [EventPageComponent],
})
export class EventModule {}
