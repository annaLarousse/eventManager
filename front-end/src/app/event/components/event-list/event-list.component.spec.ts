import { HttpClientTestingModule } from '@angular/common/http/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { lastValueFrom, of } from 'rxjs';
import { EventService } from '../../services/event.service';
import { EventListComponent } from './event-list.component';
describe('EventListComponent', () => {
  let component: EventListComponent;
  let fixture: ComponentFixture<EventListComponent>;
  let mockData: any[] = [{ test: 'test' }];
  let eventServiceMock = {
    events$: of(mockData),
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [EventListComponent],
      providers: [{ provide: EventService, useValue: eventServiceMock }],
      imports: [HttpClientTestingModule],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(EventListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should initialize events$ with the data returned from EventService', () => {
    expect(lastValueFrom(component.events$)).resolves.toEqual(mockData);
  });

  it('should display a list of all events', () => {
    expect(fixture).toMatchSnapshot();
  });
});
