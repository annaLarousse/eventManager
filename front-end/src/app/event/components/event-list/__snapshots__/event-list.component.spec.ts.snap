// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventListComponent should display a list of all events 1`] = `
<app-event-list
  __ngContext__={[Function LRootView]}
  _eventService={[Function Object]}
  displayedColumns={[Function Array]}
  events$={[Function Observable]}
>
  <h3>
    Events list
  </h3><table
    class="event-list__table mat-elevation-z8"
    mat-table=""
  >
  </table>
</app-event-list>
`;
